src/tracers/tracers.cpp

```diff
@@ -0,0 +1,293 @@
+    int rng_seed = pin->GetOrAddInteger("tracers", "initial_rng_seed", 0);
+
+    for (auto &pmb : pmesh->block_list) {
+      auto &swarm = pmb->meshblock_data.Get()->GetSwarmData()->Get("tracers");
+      // Seed is meshblock gid for consistency across MPI decomposition
+      RNGPool rng_pool(pmb->gid + rng_seed);
+
+      IndexRange ib = pmb->cellbounds.GetBoundsI(IndexDomain::interior);
+      IndexRange jb = pmb->cellbounds.GetBoundsJ(IndexDomain::interior);
+      IndexRange kb = pmb->cellbounds.GetBoundsK(IndexDomain::interior);
+
+      const auto &x_min = pmb->coords.Xf<1>(ib.s);
+      const auto &y_min = pmb->coords.Xf<2>(jb.s);
+      const auto &z_min = pmb->coords.Xf<3>(kb.s);
+      const auto &x_max = pmb->coords.Xf<1>(ib.e + 1);
+      const auto &y_max = pmb->coords.Xf<2>(jb.e + 1);
+      const auto &z_max = pmb->coords.Xf<3>(kb.e + 1);
+
+      // Create new particles and get accessor
+      auto new_particles_context = swarm->AddEmptyParticles(num_tracers_per_block);
+
+      auto &x = swarm->Get<Real>(swarm_position::x::name()).Get();
+      auto &y = swarm->Get<Real>(swarm_position::y::name()).Get();
+      auto &z = swarm->Get<Real>(swarm_position::z::name()).Get();
+      auto &id = swarm->Get<int>("id").Get();
+
+      auto swarm_d = swarm->GetDeviceContext();
+
+      const auto gid = pmb->gid;
+      pmb->par_for(
+          "SeedInitialTracers::random_per_block", 0,
+          new_particles_context.GetNewParticlesMaxIndex(),
+          KOKKOS_LAMBDA(const int new_n) {
+            auto rng_gen = rng_pool.get_state();
+            const int n = new_particles_context.GetNewParticleIndex(new_n);
+
+            x(n) = x_min + rng_gen.drand() * (x_max - x_min);
+            y(n) = y_min + rng_gen.drand() * (y_max - y_min);
+            z(n) = z_min + rng_gen.drand() * (z_max - z_min);
+            // Note that his works only during one time init.
+            // If (somehwere else) we eventually add dynamic particles, then we need to
+            // manage ids (not indices) more globally.
+            id(n) = num_tracers_per_block * gid + n;
+
+            rng_pool.free_state(rng_gen);
+
+            // TODO(pgrete) check if this actually required
+            bool on_current_mesh_block = true;
+            swarm_d.GetNeighborBlockIndex(n, x(n), y(n), z(n), on_current_mesh_block);
+          });
+    }
+  } else {
+    PARTHENON_THROW("Unknown tracer initial_seed_method");
+  }
+}
+
+TaskStatus AdvectTracers(MeshBlockData<Real> *mbd, const Real dt) {
+  auto *pmb = mbd->GetParentPointer();
+  auto &sd = pmb->meshblock_data.Get()->GetSwarmData();
+  auto &swarm = sd->Get("tracers");
+
+  auto &x = swarm->Get<Real>(swarm_position::x::name()).Get();
+  auto &y = swarm->Get<Real>(swarm_position::y::name()).Get();
+  auto &z = swarm->Get<Real>(swarm_position::z::name()).Get();
+  auto &vel_x = swarm->Get<Real>("vel_x").Get();
+  auto &vel_y = swarm->Get<Real>("vel_y").Get();
+  auto &vel_z = swarm->Get<Real>("vel_z").Get();
```

Comment:
You could reduce the memory usage by updating the particles after each RK stage. Although, that would be more complicated to implement.

Reply:
True, let's keep things simple for now (and often I imagine the velocity if the particles is also of interest).